packages/dds/map/src/mapKernel.ts

```diff
@@ -629,10 +926,28 @@ export class MapKernel {
 						pendingMessageIds[0] === localOpMetadata.pendingMessageId,
 					0x2fa /* Unexpected pending message received */,
 				);
-				pendingMessageIds.shift();
+				const pendingMessageId = pendingMessageIds.shift();
 				if (pendingMessageIds.length === 0) {
 					this.pendingKeys.delete(op.key);
 				}
+
+				if (op.type === "set") {
+					this.ackPendingSetOp(op, pendingMessageId);
+				} else if (op.type === "delete") {
+					// Adjust the keys order if it is already ack'd
+					this.decrementPendingDeleteCount(op.key);
+				}
+			} else {
+				// We do not process the remote message at this moment, but it is possible to impact the order of ack'd keys
+				if (
+					op.type === "set" &&
+					!this.pendingDeleteTracker.has(op.key) &&
+					!this.ackedKeysIndexTracker.has(op.key)
+				) {
+					this.ackedKeysIndexTracker.set(op.key, ++this.creationIndex);
+				} else if (op.type === "delete") {
+					this.ackedKeysIndexTracker.delete(op.key);
+				}
 			}
 			return false;
```

Comment:
This should always be the case, right?

Assuming so, probably better to convert this to an `else { assert(op.type === "delete"); // and the other stuff }`